network/src/commonMain/kotlin/com/wire/kalium/network/utils/NetworkUtils.kt

```diff
@@ -0,0 +1,54 @@
+package com.wire.kalium.network.utils
+
+import com.wire.kalium.network.api.ErrorResponse
+import com.wire.kalium.network.exceptions.KaliumException
+import io.ktor.client.call.*
+import io.ktor.client.features.*
+import io.ktor.client.statement.*
+import io.ktor.util.*
+import kotlinx.serialization.SerializationException
+
+sealed class NetworkResponse<out T> {
+    data class Success<out T : Any>(val response: HttpResponse, val body: T) : NetworkResponse<T>()
```

Comment:
The name of this `response` value leads to `networkResponse.successValue().response`:

```kotlin
val networkResponse = loginApi.loginWithEmail("user@example.org", "password")
networkResponse.successValue().response
```

A response inside a response.

Reply:
Also, we should not expose Ktor's internals, like `HttpResponse`.

Let's assume a cookie or header is being accessed by `Android` and `iOS`  clients:

```kotlin
val response = loginApi.loginWithEmail("user@example.org", "password")
val zuidCookie = response.successValue().response.cookie["zuid"]
```

Now, we change our dependency on Ktor to Ktor 2.0, and the cookie API changes.

Both `Android` and `iOS`, that were directly accessing Ktor in order to get the cookie, suddenly have a breaking change, that wasn't directly made by us, but introduced by us.

I suggest encapsulating whatever we want to expose by ourselves and commit to it, for example:

```kotlin
sealed class NetworkResponse<out T>(val httpStatusCode: Int, val headers: Map<String, String>) {
    
    data class Success<out T : Any>(httpStatusCode: Int, val headers: Map<String,String>, val response: HttpResponse, val body: T) : NetworkResponse<T>(httpStatusCode, headers)
    data class Error<out E : KaliumException>(httpStatusCode: Int, val headers: Map<String,String> , val kException: KaliumException) : NetworkResponse<E>(httpStatusCode, headers)
}

```

After this, we commit on supporting and making these fields (httpStatusCode and headers, in this _example_) stable and there's no way someone could access Ktor directly. If new fields are needed, then we can update the library with new features.

Reply:
I just saw the functions below with the `asHttpResponseCode()` and `responseHeaders()`, so I guess we can use those and make `response` internal.